horovod/torch/__init__.py

```diff
@@ -0,0 +1,152 @@
+            raise ValueError('named_parameters should be a sequence of '
+                             'tuples (name, parameter), usually produced by '
+                             'model.named_parameters().')
+
+        self._parameter_names = {v: k for k, v
+                                 in sorted(named_parameters)}
+        self._handles = {}
+        self._grad_accs = []
+
+        if size() > 1:
+            self._register_hooks()
+
+    def _register_hooks(self):
+        for param_group in self.param_groups:
+            for p in param_group['params']:
+                if p.requires_grad:
+                    p_tmp = p.expand_as(p)
+                    grad_acc = p_tmp.grad_fn.next_functions[0][0]
+                    grad_acc.register_hook(self._make_hook(p))
+                    self._grad_accs.append(grad_acc)
+
+    def _make_hook(self, p):
+        def hook(*ignore):
+            assert p not in self._handles
+            assert not p.grad.requires_grad
+            name = self._parameter_names.get(p)
+            handle = allreduce_async_(p.grad.data, average=True, name=name)
+            self._handles[p] = handle
+        return hook
+
+    def synchronize(self):
+        for handle in self._handles.values():
+            synchronize(handle)
+        self._handles.clear()
+
+    def step(self, closure=None):
+        self.synchronize()
+        return super(self.__class__, self).step(closure)
+
+
+def DistributedOptimizer(optimizer, named_parameters=None):
+    """
+    An optimizer that wraps another torch.optim.Optimizer, using an allreduce to
+    average gradient values before applying gradients to model weights.
+
+    Allreduce operations are executed after each gradient is computed by `loss.backward()`
+    in parallel with each other. The `step()` method ensures that all allreduce operations are
+    finished before applying gradients to the model.
+
+    DistributedOptimizer exposes the `synchronize()` method, which forces allreduce operations
+    to finish before continuing the execution. It's useful in conjunction with gradient
+    clipping, or other operations that modify gradients in place before `step()` is executed.
+
+    Example of gradient clipping:
+    ```
+    output = model(data)
+    loss = F.nll_loss(output, target)
+    loss.backward()
+    optimizer.synchronize()
+    torch.nn.utils.clip_grad_norm(model.parameters(), args.clip)
+    optimizer.step()
+    ```
+
+    Arguments:
+        optimizer: Optimizer to use for computing gradients and applying updates.
+        named_parameters: A mapping between parameter names and values. Used for naming of
+                          allreduce operations. Typically just `model.named_parameters()`.
+    """
+    # We dynamically create a new class that inherits from the optimizer that was passed in.
+    # The goal is to override the `step()` method with an allreduce implementation.
+    cls = type(optimizer.__class__.__name__, (optimizer.__class__,),
```

Comment:
It may be more readable to define _DistributedOptimizer here as a nested class. Creating a new class with the same name as the original optimizer seems a bit weird.

Reply:
I think it would be better to do this as a separate change, since Keras version have the same code structure. I'll take a note to do it later.